morpheus/_lib/src/doca/doca_context.cpp

```diff
@@ -0,0 +1,309 @@
+
+    if ((tmpu & 0xFFFFFFE0) != 0)
+    {
+        return DOCA_ERROR_INVALID_VALUE;
+    }
+
+    bdf.device = tmpu;
+
+    tmps[0] = addr[6];
+    tmps[1] = '\0';
+    tmpu    = strtoul(tmps.cbegin(), nullptr, 16);
+    if ((tmpu & 0xFFFFFFF8) != 0)
+        return DOCA_ERROR_INVALID_VALUE;
+    bdf.function = tmpu;
+
+    return DOCA_SUCCESS;
+}
+
+using jobs_check_t = doca_error_t (*)(doca_devinfo*);
+
+doca_error_t open_doca_device_with_pci(const doca_pci_bdf* value, jobs_check_t func, doca_dev** retval)
+{
+    doca_devinfo** dev_list;
+    uint32_t nb_devs;
+    doca_pci_bdf buf;
+    size_t i;
+
+    /* Set default return value */
+    *retval = nullptr;
+
+    DOCA_TRY(doca_devinfo_list_create(&dev_list, &nb_devs));
+
+    /* Search */
+    for (i = 0; i < nb_devs; i++)
+    {
+        DOCA_TRY(doca_devinfo_get_pci_addr(dev_list[i], &buf));
+        if (buf.raw == value->raw)
+        {
+            /* If any special capabilities are needed */
+            if (func != nullptr && func(dev_list[i]) != DOCA_SUCCESS)
+                continue;
+
+            /* if device can be opened */
+            auto res = doca_dev_open(dev_list[i], retval);
+            if (res == DOCA_SUCCESS)
+            {
+                doca_devinfo_list_destroy(dev_list);
+                return res;
+            }
+        }
+    }
+
+    // DOCA_LOG_WARN("Matching device not found");
+
+    doca_devinfo_list_destroy(dev_list);
+
+    return DOCA_ERROR_NOT_FOUND;
+}
+
+doca_flow_port* init_doca_flow(uint16_t port_id, uint8_t rxq_num)
+{
+    std::array<char, MAX_PORT_STR_LEN> port_id_str;
+    doca_flow_port_cfg port_cfg = {0};
+    doca_flow_port* df_port;
+    doca_flow_cfg rxq_flow_cfg = {0};
+    rte_eth_dev_info dev_info  = {0};
+    rte_eth_conf eth_conf      = {
+             .rxmode =
+            {
+                     .mtu = 2048, /* Not really used, just to initialize DPDK */
+            },
+             .txmode =
+            {
+                     .offloads = RTE_ETH_TX_OFFLOAD_IPV4_CKSUM | RTE_ETH_TX_OFFLOAD_UDP_CKSUM | RTE_ETH_TX_OFFLOAD_TCP_CKSUM,
+            },
+    };
+    rte_mempool* mp = nullptr;
+    rte_eth_txconf tx_conf;
+    rte_flow_error error;
+
+    /*
+     * DPDK should be initialized and started before DOCA Flow.
+     * DPDK doesn't start the device without, at least, one DPDK Rx queue.
+     * DOCA Flow needs to specify in advance how many Rx queues will be used by the app.
+     *
+     * Following lines of code can be considered the minimum WAR for this issue.
+     */
+
+    RTE_TRY(rte_eth_dev_info_get(port_id, &dev_info));
+    RTE_TRY(rte_eth_dev_configure(port_id, rxq_num, rxq_num, &eth_conf));
+
+    mp = rte_pktmbuf_pool_create("TEST", 8192, 0, 0, MAX_PKT_SIZE, rte_eth_dev_socket_id(port_id));
+    if (mp == nullptr)
+    {
+        // DOCA_LOG_ERR("Failed rte_pktmbuf_pool_create with: %s", rte_strerror(-ret));
+        throw std::exception();
+        // return nullptr;
```

Comment:
Left over debugging code. Also, we should be including an error message with each exception